cmd/ingest/main.go

```diff
@@ -90,117 +105,158 @@ func main() {
-			continue
+				"height": height,
+				"error":  err,
+			}).Warn("error with getblock")
+			timeout_count++
+			if timeout_count == 3 {
+				log.WithFields(logrus.Fields{
+					"timeouts":  timeout_count,
+				}).Warn("unable to issue RPC call to zcashd node 3 times")
+				timeout_count = 0
+				time.Sleep(3 * 60 * time.Second)
+			}
 		}
-
-		if len(msg) < 3 {
-			log.WithFields(logrus.Fields{
-				"msg": msg,
-			}).Warn("got unknown message type")
-			continue
+		if block != nil {
+			handleBlock(db, block)
+			height++
+			if timeout_count > 0 {
+				timeout_count--
+			}
+			//TODO store block current/prev hash for formal reorg
+		} else {
+			height -= 4
+			//TODO implement blocknotify to minimize polling on corner cases
+			time.Sleep(60 * time.Second)
 		}
+	}
+}
 
-		topic, body := msg[0], msg[1]
-
-		var sequence int
-		if len(msg[2]) == 4 {
-			sequence = int(binary.LittleEndian.Uint32(msg[len(msg)-1]))
+func getBlock(rpcClient *rpcclient.Client, height int) (*parser.Block, error) {
+	params := make([]json.RawMessage, 2)
+	params[0] = json.RawMessage("\"" + strconv.Itoa(height) + "\"")
+	params[1] = json.RawMessage("0")
+	result, rpcErr := rpcClient.RawRequest("getblock", params)
+
+	var err error
+	var errCode int64
+
+	// For some reason, the error responses are not JSON
+	if rpcErr != nil {
+		errParts := strings.SplitN(rpcErr.Error(), ":", 2)
+		errCode, err = strconv.ParseInt(errParts[0], 10, 32)
+		//Check to see if we are requesting a height the zcashd doesn't have yet
+		if err == nil && errCode == -8 {
+			return nil, nil
 		}
+		return nil, errors.Wrap(rpcErr, "error requesting block")
+	}
 
-		switch string(topic) {
-
-		case opts.zmqTopic:
-			handleBlock(db, sequence, body)
+	var blockDataHex string
+	err = json.Unmarshal(result, &blockDataHex)
+	if err != nil{
+		return nil, errors.Wrap(err, "error reading JSON response")
+	}
 
-		default:
-			log.WithFields(logrus.Fields{
-				"seqnum": sequence,
-				"topic":  topic,
-			}).Warn("got message with unknown topic")
-		}
+	blockData, err := hex.DecodeString(blockDataHex)
+	if err != nil {
+		return nil, errors.Wrap(err, "error decoding getblock output")
 	}
-}
 
-func handleBlock(db *sql.DB, sequence int, blockData []byte) {
 	block := parser.NewBlock()
 	rest, err := block.ParseFromSlice(blockData)
 	if err != nil {
-		log.WithFields(logrus.Fields{
-			"seqnum": sequence,
-			"error":  err,
-		}).Error("error parsing block")
-		return
+		return nil, errors.Wrap(err, "error parsing block")
 	}
 	if len(rest) != 0 {
+		return nil, errors.New("received overlong message")
+	}
+	return block, nil
+}
+
+
+func handleBlock(db *sql.DB, block *parser.Block) {
+	nodeBlockHash := hex.EncodeToString(block.GetEncodableHash())
+
+	nodeBlockHeight := block.GetHeight()
+	localBlockHash, err := storage.GetBlockHash(db, nodeBlockHeight)
+
+	// if local block has different hash delete it before store new block
+	if localBlockHash != "" && localBlockHash != nodeBlockHash {
```

Comment:
It appears that this, effectively, does an insert-replace for new blocks. That would fix single-block reorgs but it does not address larger ones. Given the change below to also [rewind the server by 4 blocks](https://github.com/zcash-hackworks/lightwalletd/pull/56/files#diff-0831e2c5d1d4855c0db87d71286f96a6R172) every time it reaches the chain tip, this would catch a lot of small reorgs but would fail on larger ones.

In other words, it seems these changes reduce the reorg problem but we could go a bit further and completely fix it (for instance by validating the prevHashes to confirm that every block is on-chain and the rewinding until we find the beginning of the chain split).